core/fetcher.go

```diff
@@ -97,45 +159,55 @@ func (f *BlockFetcher) Commit(ctx context.Context, height *int64) (*types.Commit
 		return nil, fmt.Errorf("core/fetcher: commit not found at height %d", height)
 	}
 
-	return res.Commit, nil
+	commit, err := types.CommitFromProto(res.Commit)
+	if err != nil {
+		return nil, err
+	}
+
+	return commit, nil
 }
 
 // ValidatorSet queries Core for the ValidatorSet from the
 // block at the given height.
-func (f *BlockFetcher) ValidatorSet(ctx context.Context, height *int64) (*types.ValidatorSet, error) {
-	perPage := 100
-
-	vals, total := make([]*types.Validator, 0), -1
-	for page := 1; len(vals) != total; page++ {
-		res, err := f.client.Validators(ctx, height, &page, &perPage)
-		if err != nil {
-			return nil, err
-		}
+// If the height is nil, use the latest height.
+func (f *BlockFetcher) ValidatorSet(ctx context.Context, height int64) (*types.ValidatorSet, error) {
+	// return error if the client is still not started
+	if !f.client.IsRunning() {
+		return nil, errors.New("client not running")
+	}
+	res, err := f.client.ValidatorSet(ctx, &coregrpc.ValidatorSetRequest{Height: height})
+	if err != nil {
+		return nil, err
+	}
 
-		if res != nil && len(res.Validators) == 0 {
-			return nil, fmt.Errorf("core/fetcher: validator set not found at height %d", height)
-		}
+	if res != nil && res.ValidatorSet == nil {
+		return nil, fmt.Errorf("core/fetcher: validator set not found at height %d", height)
+	}
 
-		total = res.Total
-		vals = append(vals, res.Validators...)
+	validatorSet, err := types.ValidatorSetFromProto(res.ValidatorSet)
+	if err != nil {
+		return nil, err
 	}
 
-	return types.NewValidatorSet(vals), nil
+	return validatorSet, nil
 }
 
 // SubscribeNewBlockEvent subscribes to new block events from Core, returning
 // a new block event channel on success.
 func (f *BlockFetcher) SubscribeNewBlockEvent(ctx context.Context) (<-chan types.EventDataSignedBlock, error) {
-	// start the client if not started yet
+	// return error if the client is still not started
 	if !f.client.IsRunning() {
 		return nil, errors.New("client not running")
 	}
-
+	if f.isListeningForBlocks {
```

Comment:
Feels like we have a race condition here. WDYT to change `isListeningForBlocks` to `atomic.Bool` ?

Reply:
504591866243ad03a246957f8d61182192c1e5c2